internal/dag/cache.go

```diff
@@ -74,6 +74,27 @@ func (kc *KubernetesCache) matchesIngressClass(obj Object) bool {
 	objectClass := ingressClass(obj)
 	targetClass := stringOrDefault(kc.IngressClass, DEFAULT_INGRESS_CLASS)
 
+	// If an ingressClass is defined, match only that
+	if kc.IngressClass != "" {
+		switch objectClass {
+		case kc.IngressClass:
+			return true
+		// Any other ingress class fails to match.
+		default:
+			kind := k8s.KindOf(obj)
+			om := obj.GetObjectMeta()
+
+			kc.WithField("name", om.GetName()).
+				WithField("namespace", om.GetNamespace()).
+				WithField("kind", kind).
+				WithField("ingress.class", objectClass).
+				Debug("ignoring object with unmatched ingress class")
+
+			return false
+		}
+	}
+
+	// User defined class not specified
```

Comment:
I think that we can simplify the logic to this:

```Go
       objectClass := ingressClass(obj)

        switch objectClass {
        case kc.IngressClass:
                // Handles kc.IngressClass == "" and kc.IngressClass == "custom".
                return true
        case DEFAULT_INGRESS_CLASS:
                // kc.IngressClass == "" implicitly matches the default too.
                if kc.IngressClass == "" {
                        return true
                }
        }

        // Any other ingress class fails to match.
        kind := k8s.KindOf(obj)
        om := obj.GetObjectMeta()

        kc.WithField("name", om.GetName()).
                WithField("namespace", om.GetNamespace()).
                WithField("kind", kind).
                WithField("ingress.class", objectClass).
                Debug("ignoring object with unmatched ingress class")

        return false
```

If there's an exact match, then we have a match. The only other case we need to worry about is when Contour is running with an empty ingress class and we want to implicitly match the default.

Reply:
I agree with James here, this is a much clearer function to read, and it means we don't make the unneeded targetClass variable.

From what I can see, the available states are:
- [x] no configured ingress.class, none on object - pass
- [x] no configured ingress.class, "contour" on object - pass
- [x] no configured ingress.class, anything else on object - fail
- [x] configured ingress.class, none on object - fail
- [x] configured ingress.class, correct on object
- [x] configured ingress.class, incorrect on object

I don't think we currently do enough testing of the "no configured ingress class" cases.